src/telemetry_poller.erl

```diff
@@ -0,0 +1,407 @@
+start_link(Opts) when is_list(Opts) ->
+    Name = proplists:get_value(name, Opts, ?MODULE),
+    Args = parse_args(Opts),
+    gen_server:start_link({local, Name}, ?MODULE, Args, []).
+
+%% @doc
+%% Stops the poller with specified reason.
+%%
+%% See documentation for {@link gen_server:stop/3} to learn more about the behaviour of this function.
+-spec stop(t()) -> ok.
+stop(Poller) ->
+    stop(Poller, normal, inifinity).
+
+-spec stop(t(), term()) -> ok.
+stop(Poller, Reason) ->
+    stop(Poller, Reason, inifinity).
+
+-spec stop(t(), term(), timeout()) -> ok.
+stop(Poller, Reason, Timeout) ->
+    gen_server:stop(Poller, Reason, Timeout).
+
+%% @doc
+%% Returns a list of measurements used by the poller.
+-spec list_measurements(t()) -> [measurement()].
+list_measurements(Poller) ->
+    gen_server:call(Poller, get_measurements).
+
+-spec init(map()) -> {ok, state()}.
+init(Args) ->
+    schedule_measurement(0),
+    {ok,
+     #{measurements => maps:get(measurements, Args),
+       period => maps:get(period, Args)}}.
+
+%% @doc
+%% Returns a child spec for the poller for running under a supervisor.
+child_spec(Opts) ->
+    Id =
+        case proplists:get_value(name, Opts, telemetry_poller_default) of
+            Name when is_atom(Name) -> Name;
+            {global, Name} -> Name;
+            {via, _, Name} -> Name
+        end,
+
+    #{
+        id => Id,
+        start => {telemetry_poller, start_link, [Opts]}
+    }.
+
+parse_args(Args) ->
+    Measurements = proplists:get_value(measurements, Args, []),
+    validate_measurements(Measurements),
+    Period = proplists:get_value(period, Args, 5000),
+    validate_period(Period),
+    VMMeasurements = parse_vm_measurements(proplists:get_value(vm_measurements, Args, [])),
+
+    #{measurements => Measurements ++ VMMeasurements, period => Period}.
+
+-spec parse_vm_measurements([] | default) -> [vm_measurement()].
+parse_vm_measurements([]) ->
+    [];
+parse_vm_measurements(default) ->
+    parse_vm_measurements(?DEFAULT_VM_MEASUREMENTS);
+parse_vm_measurements(Measurements) ->
+    __Measurements = sets:from_list(Measurements),
+    lists:map(fun parse_vm_measurement/1, sets:to_list(__Measurements)).
+
+-spec parse_vm_measurement(measurement()) -> {telemetry_poller_vm, measurement(), list()} | no_return().
+parse_vm_measurement(Measurement) ->
+    case lists:member(Measurement, ?DEFAULT_VM_MEASUREMENTS) of
+        true -> vm_measurement(Measurement);
+        false -> erlang:error(badarg, Measurement)
+    end.
+
+-spec vm_measurement(atom()) -> measurement().
+vm_measurement(Function) ->
+    {telemetry_poller_vm, Function, []}.
+
+-spec schedule_measurement(non_neg_integer()) -> ok.
+schedule_measurement(CollectInMillis) ->
+    erlang:send_after(CollectInMillis, self(), collect), ok.
+
+-spec validate_period(term()) -> ok | no_return().
+validate_period(Period) when is_integer(Period), Period > 0 ->
+    ok;
+validate_period(Term) ->
+    erlang:error(badarg, Term).
```

Comment:
Let's please raise nice error messages here and below. :) Also note that the second argument to `erlang:error` has to be a list. So I recommend writing some tests for those failures and make sure the error messages are understandable by humans. :)

```suggestion
    erlang:error({badarg, "expected period to be a positive integer"}, [Term]).
```

Reply:
Another alternative is to LOG_ERROR and then error with badarg.